stats/opentelemetry/internal/grpc_trace_bin_propagator_test.go

```diff
@@ -42,151 +42,195 @@ func Test(t *testing.T) {
+// For slow path, it passes `otel.MapCarrier` as carrier to `Inject()` which
+// injects the span context after base64 encoding as string using
+// `carrier.Set()`. It then retrieves the injected span context using
+// `carrier.Get()` for verification because `Set()` does injection in string
+// format.
+func (s) TestInject(t *testing.T) {
+	tests := []struct {
+		name string
+		sc   oteltrace.SpanContext
+		fast bool
+	}{
+		{
+			name: "fast path, valid context",
+			sc: oteltrace.SpanContext{}.WithTraceID(
+				oteltrace.TraceID{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16}).WithSpanID(
+				oteltrace.SpanID{17, 18, 19, 20, 21, 22, 23, 24}).WithTraceFlags(
+				oteltrace.TraceFlags(1)),
+			fast: true,
+		},
+		{
+			name: "fast path, invalid context",
+			sc:   oteltrace.SpanContext{},
+			fast: true,
+		},
+		{
+			name: "slow path, valid context",
+			sc: oteltrace.SpanContext{}.WithTraceID(
+				oteltrace.TraceID{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16}).WithSpanID(
+				oteltrace.SpanID{17, 18, 19, 20, 21, 22, 23, 24}).WithTraceFlags(
+				oteltrace.TraceFlags(1)),
+			fast: false,
+		},
+		{
+			name: "slow path, invalid context",
+			sc:   oteltrace.SpanContext{},
+			fast: false,
+		},
 	}
-}
 
-// TestExtract_FastPath verifies that the GRPCTraceBinPropagator correctly
-// extracts OpenTelemetry trace context data using the CustomCarrier.
-//
-// It is called the fast path because it extracts the trace context directly
-// in the binary format.
-func (s) TestExtract_FastPath(t *testing.T) {
-	p := GRPCTraceBinPropagator{}
-	sc := oteltrace.NewSpanContext(oteltrace.SpanContextConfig{
-		TraceID:    [16]byte{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16},
-		SpanID:     [8]byte{17, 18, 19, 20, 21, 22, 23, 24},
-		TraceFlags: oteltrace.FlagsSampled,
-		Remote:     true,
-	})
-	bd := binary(sc)
-
-	ctx, cancel := context.WithCancel(context.Background())
-	defer cancel()
-	c := itracing.NewCustomCarrier(stats.SetIncomingTrace(ctx, bd))
-	tCtx := p.Extract(ctx, c)
-	got := oteltrace.SpanContextFromContext(tCtx)
-
-	if !got.Equal(sc) {
-		t.Fatalf("got = %v, want %v", got, sc)
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			p := GRPCTraceBinPropagator{}
+			tCtx, tCancel := context.WithCancel(context.Background())
+			tCtx = oteltrace.ContextWithSpanContext(tCtx, test.sc)
+			defer tCancel()
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+
+			var c otelpropagation.TextMapCarrier
+			if test.fast {
+				c = itracing.NewCustomCarrier(metadata.NewOutgoingContext(ctx, metadata.MD{}))
+			} else {
+				c = otelpropagation.MapCarrier{}
+			}
+			p.Inject(tCtx, c)
+			var got oteltrace.SpanContext
+			if test.fast {
+				got, _ = fromBinary(stats.OutgoingTrace(c.(*itracing.CustomCarrier).Context()))
+			} else {
+				b, _ := base64.StdEncoding.DecodeString(c.Get(itracing.GRPCTraceBinHeaderKey))
+				got, _ = fromBinary(b)
```

Comment:
Let's not get in the habit of ignoring errors, it's rare we don't ignore errors in test or implementation. Fail the test if it does error.

Reply:
for `fromBinary` ignored values are just booleans indicating success or not. So eventual comparison of want and got would have failed anyway in case want was non zero. Anyways i have handled them separately now. It required some extra variable declarations and if conditions.